Scala/spectral-norm/spectralnorm.scala

```diff
@@ -0,0 +1,85 @@
+/* The Computer Language Benchmarks Game
+   http://benchmarksgame.alioth.debian.org/
+   contributed by Isaac Gouy
+   parallel by the Anh Team
+   Scala Futures version by Robert Wilton
+ */
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.annotation.tailrec
+import ExecutionContext.Implicits.global
+
+object spectralnorm {
+  // Ordinary and transposed versions of infinite matrix
+  val A = (x: Int, y: Int) => 1.0 / ((x + y) * (x + y + 1) / 2 + x + 1)
+  val At = (x: Int, y: Int) => 1.0 / ((y + x) * (y + x + 1) / 2 + y + 1)
+
+  case class Ctx(n: Int) {
+    val u, v, tmp = Array.fill(n)(1.0)
+    private implicit val ctx: Ctx = this
+    val multiplyAv = multiply(A) _
+    val multiplyAtv = multiply(At) _
+  }
+
+  def main(args: Array[String]): Unit = {
+    val n = if (args.length > 0) args(0).toInt else 100
+    implicit val ctx: Ctx = Ctx(n)
+
+    // Calculate the chunks and perform calculation.
+    val chunks = {
+      val threads = Runtime.getRuntime.availableProcessors
+      val size = 1 + n / threads
+      for {
+        threadId <- 0 until threads
+        start = threadId * size
+        end = ((threadId + 1) * size).min(n)
+      } yield start until end
```

Comment:
To be honest, I think it would be more efficient to just let the `ExecutionContext` handle that part (splitting the work into parallel chunks). This is just adding overhead (both for readability and performance).

Reply:
I agree about that in case of idiomatic test, but in non-idiomatic we should still use calculated chunk - or else we would observe much slower execution - on my machine it was ~3.5s with explicit chunks vs ~6s with spawning N Futures. It might be better if we would use parallel collections, but they're not a part of standard library anymore. 
The main problem with leaving whole shcheduling to ExecutionContext is large overhead created by context switches and actual running N threads. In case of manual chunks there should be almost no context switches as the number of Futures is equal to number of available CPU cores.

Reply:
Oh, that’s an interesting result, and not intuitive to me!

I would not expect the default execution context to create N actual threads (unless we use `concurrent.blocking`?). By default, it should be bounded by the number of processors.